core/client/src/main/java/alluxio/client/block/stream/NettyPacketReader.java

```diff
@@ -0,0 +1,410 @@
+        mChannel.close();
+        throw Throwables.propagate(e);
+      }
+
+      while (true) {
+        try {
+          ByteBuf buf = readPacket();
+          // A null packet indicates the end of the stream.
+          if (buf == null) {
+            return;
+          }
+          buf.release();
+        } catch (IOException e) {
+          LOG.warn("Failed to close the NettyBlockReader (block: {}, address: {}).",
+              mId, mAddress, e);
+          try {
+            mChannel.close().sync();
+          } catch (InterruptedException ee) {
+            throw Throwables.propagate(ee);
+          }
+          return;
+        }
+      }
+    } finally {
+      if (mChannel.isOpen()) {
+        Preconditions.checkState(mChannel.pipeline().last() instanceof PacketReadHandler);
+        mChannel.pipeline().removeLast();
+
+        // Make sure "autoread" is on before realsing the channel.
+        resume();
+      }
+      mContext.releaseNettyChannel(mAddress, mChannel);
+      mClosed = true;
+    }
+  }
+
+  /**
+   * @return bytes remaining
+   */
+  private long remaining() {
+    return mStart + mBytesToRead - mPosToRead;
+  }
+
+  /**
+   * @return true if there are too many packets pending
+   */
+  private boolean tooManyPacketsPending() {
+    return mPackets.size() >= MAX_PACKETS_IN_FLIGHT;
+  }
+
+  /**
+   * The netty handler that reads packets from the channel.
+   */
+  private class PacketReadHandler extends ChannelInboundHandlerAdapter {
+    /**
+     * Default constructor.
+     */
+    public PacketReadHandler() {}
+
+    @Override
+    public void channelRead(ChannelHandlerContext ctx, Object msg) throws IOException {
+      Preconditions.checkState(acceptMessage(msg), "Incorrect response type %s, %s.",
+          msg.getClass().getCanonicalName(), msg);
+
+      RPCProtoMessage response = (RPCProtoMessage) msg;
+      Protocol.Status status = ((Protocol.Response) response.getMessage()).getStatus();
+      if (!Status.isOk(status)) {
+        throw new IOException(String
+            .format("Failed to read block %d from %s with status %s.", mId, mAddress,
+                status.toString()));
+      }
+      mLock.lock();
+      try {
+        Preconditions.checkState(mPacketReaderException == null);
+        DataBuffer dataBuffer = response.getPayloadDataBuffer();
+        ByteBuf buf;
+        if (dataBuffer == null) {
+          buf = ctx.alloc().buffer(0, 0);
+        } else {
+          Preconditions.checkState(dataBuffer.getLength() > 0);
+          assert dataBuffer.getNettyOutput() instanceof ByteBuf;
+          buf = (ByteBuf) dataBuffer.getNettyOutput();
+        }
+        Preconditions.checkState(mPackets.offer(buf));
+        mNotEmptyOrFailed.signal();
+
+        if (tooManyPacketsPending()) {
+          pause();
+        }
+      } catch (Throwable e) {
+        if (response.getPayloadDataBuffer() != null) {
+          response.getPayloadDataBuffer().release();
+        }
+        throw e;
```

Comment:
i've updated the code and removed this exception handling. It should be clearer.